client/client.go

```diff
@@ -48,6 +48,23 @@ var DefaultTransport CancelableTransport = &http.Transport{
 	TLSHandshakeTimeout: 10 * time.Second,
 }
 
+type EndpointSelectionMode int
+
+const (
+	// EndpointSelectionDefault is to pick an endpoint in a random manner.
+	EndpointSelectionDefault EndpointSelectionMode = iota
+
+	// EndpointSelectionPrioritizeLeader is to prioritize leader for reducing needless
+	// forward between follower and leader.
+	//
+	// This mode should be used with Client.AutoSync().
+	EndpointSelectionPrioritizeLeader
+
+	// EndpointSelectionPreference is to manually picks an endpoint
+	// with using Config.PreferenceIndex as an index.
+	EndpointSelectionPreference
```

Comment:
If we keep enum, I suggest to keep the below switch for easy extensibility. We just need to add a new case for adding a new mode. How do you think?

And you mentioned about the idea of local and consistent read ( https://github.com/coreos/etcd/pull/3950#issuecomment-165341542 ). It is the original motivation of the preference mode. Will the feature not appear soon?